org.testeditor.web.dropwizard.xtext/src/main/java/org/testeditor/web/dropwizard/xtext/validation/ValidationMarkerUpdater.xtend

```diff
@@ -0,0 +1,110 @@
+ * as a callback invoked after validating each resource.
+ * 
+ * This updater in turn sends the summarized validation results to a
+ * {@link org.testeditor.web.dropwizard.xtext.validation.ValidationMarkerMap},
+ * but only after the builder has finished validating all resources. Since the
+ * latter does not offer appropriate callback hooks, 
+ * {@link org.testeditor.web.xtext.index.CustomStandaloneBuilder} extends that
+ * class, and calls this updater's {@link #updateMarkerMap) after the build is
+ * complete.
+ * 
+ * While the mode of usage by the builder means that all issues passed to 
+ * {@link #handleIssue} will refer to the same resource, the method is, in
+ * theory, able to handle a mixed set of issues belonging to arbitrary
+ * resources. Unfortunately, not all validators produce issues with a non-null
+ * resource URI set, so as a fallback, the context resource can be explicitly
+ * set with {@link #setContext}, and its URI will be used to construct a
+ * resource path string for any issue that lacks the necessary information
+ * itself. 
+ */
+@Singleton
+class ValidationMarkerUpdater implements IIssueHandler {
+
+	@Inject DefaultIssueHandler delegateForDefaultBehavior
+	@Inject extension ValidationMarkerMap validationMarkerMap
+
+	static val logger = LoggerFactory.getLogger(ValidationMarkerUpdater)
+
+	val collectedValidationSummaries = new HashMap<String, ValidationSummary>()
+	var Path rootDirectory
+	var Resource currentResource
+
+	def void init(String rootDirectory) {
+		this.rootDirectory = Paths.get(rootDirectory)
+	}
+
+	def setContext(Resource resource) {
+		currentResource = resource
+	}
+
+	override handleIssue(Iterable<Issue> issues) {
+		if (issues !== null && issues.length > 0) {
+			issues.forEach [
+				val path = uriToProblem.resourcePath
+				if (path !== null) {
+					previouslyCollectedSummary(path).incrementFor(it)
+				} else {
+					logger.warn('''Could not determine resource path for issue: «it»''')
+				}
+			]
+		}
+		return delegateForDefaultBehavior.handleIssue(issues)
+	}
+
+	def updateMarkerMap() {
+		if (!collectedValidationSummaries.empty) {
+			updateMarkers(new ArrayList(collectedValidationSummaries.values))
+			collectedValidationSummaries.clear
+		}
+	}
+
+	private def previouslyCollectedSummary(String path) {
+		collectedValidationSummaries.computeIfAbsent(path, [ValidationSummary.noMarkers(path)])
+	}
+
+	private def incrementFor(ValidationSummary summary, Issue issue) {
+		summary => [
+			switch (issue.severity) {
+				case ERROR:
+					errors = errors + 1
+				case WARNING:
+					warnings = warnings + 1
+				case INFO:
+					infos = infos + 1
+				case IGNORE: {
+				}
+			}
+		]
+	}
+
+	private def getResourcePath(URI uri) {
+		if (uri !== null) {
+			return rootDirectory.relativize(Paths.get(uri.path)).toString
+		} else if (currentResource !== null) {
+			return rootDirectory.relativize(Paths.get(currentResource.URI.path)).toString
+		} else {
+			return null
```

Comment:
does this make sense, or should fail early apply here

Reply:
`handleIssue`, the only method using this, needs to be resilient enough to skip an issue for which the resource / path cannot be determined. That way, it still provides some service, as opposed to none at all, in the face of errors (graceful degradation). So, I _do not_ want this method to throw an exception in case it cannot determine the path (and no other issues arose). If it did, I'd have to catch it in `handleIssue`, and I don't think throwing an exception from a private method only to catch it in another method of the same class makes a lot of sense.


Side note: I initially did not ensure that `rootDirectory` would always be representing an _absolute_ path. If it is not, `relativize` will throw an exception when called with a path that actually _is_ absolute. If this occurs during startup, it takes down the whole application (crashes it), even though I feel like this is a problem that can be recovered from (by providing a potentially degraded service).